extensions/aws/processors/PutS3Object.cpp

```diff
@@ -186,6 +250,48 @@ void PutS3Object::setAttributes(
   }
 }
 
+void PutS3Object::ageOffMultipartUploads(const CommonProperties &common_properties) {
+  const auto now = std::chrono::system_clock::now();
+  if (now - last_ageoff_time_.load() < multipart_upload_ageoff_interval_) {
+    logger_->log_debug("Multipart Upload Age off interval still in progress, not checking obsolete multipart uploads.");
+    return;
+  }
+
+  logger_->log_trace("Listing aged off multipart uploads still in progress.");
+  aws::s3::ListMultipartUploadsRequestParameters list_params(common_properties.credentials, *client_config_);
+  list_params.setClientConfig(common_properties.proxy, common_properties.endpoint_override_url);
+  list_params.bucket = common_properties.bucket;
+  list_params.age_off_limit = multipart_upload_max_age_threshold_;
+  list_params.use_virtual_addressing = use_virtual_addressing_;
+  auto aged_off_uploads_in_progress = s3_wrapper_.listMultipartUploads(list_params);
+  if (!aged_off_uploads_in_progress) {
+    logger_->log_error("Listing aged off multipart uploads failed!");
+    return;
+  }
+
+  logger_->log_info("Found %d aged off pending multipart upload jobs in bucket '%s'", aged_off_uploads_in_progress->size(), common_properties.bucket);
+  size_t aborted = 0;
+  for (const auto& upload : *aged_off_uploads_in_progress) {
+    logger_->log_info("Aborting multipart upload with key '%s' and upload id '%s' in bucket '%s'", upload.key, upload.upload_id, common_properties.bucket);
+    aws::s3::AbortMultipartUploadRequestParameters abort_params(common_properties.credentials, *client_config_);
+    abort_params.setClientConfig(common_properties.proxy, common_properties.endpoint_override_url);
+    abort_params.bucket = common_properties.bucket;
+    abort_params.key = upload.key;
+    abort_params.upload_id = upload.upload_id;
+    abort_params.use_virtual_addressing = use_virtual_addressing_;
+    if (!s3_wrapper_.abortMultipartUpload(abort_params)) {
+       logger_->log_error("Failed to abort multipart upload with key '%s' and upload id '%s' in bucket '%s'", abort_params.key, abort_params.upload_id, abort_params.bucket);
+       continue;
+    }
+    ++aborted;
+  }
+  if (aborted > 0) {
+    logger_->log_info("Aborted %d pending multipart upload jobs in bucket '%s'", aborted, common_properties.bucket);
+  }
+  s3_wrapper_.ageOffLocalS3MultipartUploadStates(multipart_upload_max_age_threshold_);
+  last_ageoff_time_ = now;
```

Comment:
I didn't think it was necessary as in scenario where there are overlapping onTriggers the worst case scenario would be that the aged off uploads are requested to be deleted multiple times. Of course it could be improved to avoid this scenario, I didn't see how this could be done with a single atomic `compare_and_exchange` call, so I removed the atomic variable and added a mutex for this block instead in bd5d6d1b50c528061ffc34529e213039bd6dbf7a

Reply:
I didn't do too well on phrasing this comment, but my concern was updating `last_ageoff_time_` at the end, instead of updating early and atomically/locked, then proceeding with the work on the current thread. I think it could more easily happen with the old version that multiple threads start deleting the same aged off uploads. The current version avoids this, even though it doesn't seem critical.